plutus-contract/src/Language/Plutus/Contract/Wallet.hs

```diff
@@ -124,12 +130,28 @@ addOutputs pk vl tx = tx & over Tx.outputs (pko :) where
 
 -- | Balance an unabalanced transaction, sign it, and submit
 --   it to the chain in the context of a wallet.
-handleTx :: MonadWallet m => UnbalancedTx -> m Tx
-handleTx utx =
-    balanceWallet utx >>= addSignatures (view T.requiredSignatures utx) >>= WAPI.signTxAndSubmit
-
--- | Add signatures of the private keys belonging to the given
---   public keys to the transaction, provided that they are
---   known to the wallet.
-addSignatures :: MonadWallet m => [PubKey] -> Tx -> m Tx
-addSignatures pks tx = foldM WAPI.signTxnWithKey tx pks
+handleTx :: MonadWallet m => SigningProcess -> UnbalancedTx -> m Tx
+handleTx p utx =
+    balanceWallet utx >>= addSignatures p (view T.requiredSignatures utx) >>= WAPI.signTxAndSubmit
+
+-- | The signing process gets a finished transaction and a list of public keys,
+--   and signs the transaction with the corresponding private keys.
+newtype SigningProcess = SigningProcess { addSignatures :: forall m. MonadWallet m => [PubKey] -> Tx -> m Tx }
```

Comment:
Looks like an effect handler to me :laughing:

Reply:
It totally is an effect handler :D